domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,202 @@
+	token_rotate_interval AS &SecretBackend.token_rotate_interval
+FROM secret_backend
+WHERE uuid = $M.uuid`, SecretBackend{}, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertBackendStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend
+    (uuid, name, backend_type, token_rotate_interval)
+VALUES (
+    $SecretBackend.uuid,
+    $SecretBackend.name,
+    $SecretBackend.backend_type,
+    $SecretBackend.token_rotate_interval
+)
+ON CONFLICT (uuid) DO UPDATE SET
+    name=EXCLUDED.name,
+    token_rotate_interval=EXCLUDED.token_rotate_interval;`, SecretBackend{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertRotationStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_rotation
+    (backend_uuid, next_rotation_time)
+VALUES (
+    $SecretBackendRotation.backend_uuid,
+    $SecretBackendRotation.next_rotation_time
+)
+ON CONFLICT (backend_uuid) DO UPDATE SET next_rotation_time=EXCLUDED.next_rotation_time;`,
+		SecretBackendRotation{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.clearConfigStmt, err = sqlair.Prepare(`
+DELETE FROM secret_backend_config
+WHERE backend_uuid = $M.uuid AND name NOT IN ($S[:]);`,
+		sqlair.M{}, sqlair.S{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertConfigStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_config
+    (backend_uuid, name, content)
+VALUES (
+    $SecretBackendConfig.backend_uuid,
+    $SecretBackendConfig.name,
+    $SecretBackendConfig.content
+)
+ON CONFLICT (backend_uuid, name) DO UPDATE SET content = EXCLUDED.content`,
+		SecretBackendConfig{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (o *upsertOperation) validate(ctx context.Context, tx *sqlair.TX) error {
+	var existing SecretBackend
+	err := tx.Query(ctx, o.getBackendStmt, sqlair.M{"uuid": o.ID}).Get(&existing)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		// New insert.
+		if o.Name == "" {
+			return fmt.Errorf("backend name is missing: %w", backenderrors.NotValid)
+		}
+		if o.BackendType == "" {
+			return fmt.Errorf("backend type is missing: %w", backenderrors.NotValid)
+		}
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	// Update.
+	if existing.BackendType == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend type is empty for backend %q", o.ID)
+	}
+	if existing.Name == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend name is empty for backend %q", o.ID)
+	}
+	if o.BackendType != "" && o.BackendType != existing.BackendType {
+		// The secret backend type is immutable.
+		return errors.Errorf(
+			"cannot change backend type from %q to %q because backend type is immutable",
+			existing.BackendType, o.BackendType,
+		)
+	}
+	// Fill in the existing backend type.
+	o.BackendType = existing.BackendType
+	if o.Name == "" {
+		// Fill in the existing name.
+		o.Name = existing.Name
```

Comment:
validate should not mutate any data - it should just error if there's an issue
We can fill in any data during the Prepare() step

Reply:
The whole point here is to ensure Prepare() only for building statements without db access.
And the reason why we need to fill the `Name`, `TokenRotateInterval` etc. is to allow you only provide the data you want to update. If we don't fill it here, we have to use condition check in UPDATE SET section(which makes the sql to complex).

```
ON CONFLICT(id) DO UPDATE SET
-- also we have to handle NULL input probably
name = CASE WHEN excluded.name != '' THEN excluded.name ELSE name END,
token_rotate_interval = CASE WHEN excluded.token_rotate_interval != '' THEN excluded.token_rotate_interval ELSE token_rotate_interval END;
```

So I think the logic here is correct, but I agree the `validate` is not a good name.
I consider to move the filling data part into `Apply` but only do validation in `validate`.